adaptive/learner/triangulation.py

```diff
@@ -1,20 +1,46 @@
-import math
 from collections import Counter
 from collections.abc import Iterable, Sized
 from itertools import chain, combinations
 from math import factorial
-
-import numpy as np
 import scipy.spatial
 
+from numpy import (
+    square,
+    zeros,
+    subtract,
+    array,
+    ones,
+    dot,
+    asarray,
+    concatenate,
+    average,
+    eye,
+    mean,
+    abs,
+    sqrt,
+)
+from numpy import sum as nsum
+from numpy.linalg import det as ndet
+from numpy.linalg import slogdet, solve, matrix_rank, norm
+
 
 def fast_norm(v):
+    """ Manually take the vector norm for len 2, 3 vectors. Defaults to a square root of the dot product
+    for larger vectors.
+
+    Note that for large vectors, it is possible for integer overflow to occur.
+    For instance:
+    vec = [49024, 59454, 12599, -63721, 18517, 27961]
+    dot(vec, vec) = -1602973744
+
+    """
+    len_v = len(v)
     # notice this method can be even more optimised
-    if len(v) == 2:
-        return math.sqrt(v[0] * v[0] + v[1] * v[1])
-    if len(v) == 3:
-        return math.sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2])
-    return math.sqrt(np.dot(v, v))
+    if len_v == 2:
+        return sqrt(v[0] * v[0] + v[1] * v[1])
+    if len_v == 3:
+        return sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2])
+    return sqrt(dot(v, v))
```

Comment:
It seems like sqrt from the `math` library is much faster on floats than `numpy`:
```
In [1]: from numpy import sqrt as ns

In [2]: from math import sqrt as ms

In [3]: %timeit ns(0.32)
705 ns ± 14.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit ms(0.32)
46 ns ± 0.362 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```